### PR TITLE
Initial fail2ban role

### PR DESCRIPTION
**Pull request by adriaan-niess**
Implementation of fail2ban role

### CODE REVIEW

File: defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+---
+fail2ban_bantime: 3600
+fail2ban_findtime: 600
+fail2ban_maxretry: 3
+fail2ban_enabled: "false"
```
Comment (SlothOfAnarchy) [inline defaults/main.yml]: Please use a boolean
Either jinja directly uses true/false when specifying it as boolean, otherwise you can add an if in the template to check the boolean value and write the string accordingly

File: defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+fail2ban_backend: systemd
+fail2ban_logencoding: auto
+fail2ban_ignoreip:
+  - "127.0.0.1/8"
+fail2ban_additional_filters: []
+fail2ban_jails:
```
Comment (SlothOfAnarchy) [inline defaults/main.yml:10]: If every entry has a name, there are no duplicates and there needs to be no way of unsetting an entry within our config repo, you could use a dict for this.

File: defaults/main.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  - name: nginx-limit-req
+    port: http,https
+    logpath: "%(nginx_error_log)s"
+  - name: nginx-http-auth
+    port: http,https
+    logpath: "%(nginx_error_log)s"
+  - name: nginx-botsearch
+    port: http,https
+    logpath: "%(nginx_error_log)s"
+    maxretry: 2
```
Comment (SlothOfAnarchy) [inline defaults/main.yml]: missing newline at the end of the file

File: handlers/main.yml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+- name: Restart fail2ban
+  service:
+    name: fail2ban
+    state: reloaded
```
Comment (SlothOfAnarchy) [inline handlers/main.yml]: missing newline at the end of the file

File: handlers/main.yml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+- name: Restart fail2ban
```
Comment (SlothOfAnarchy) [inline handlers/main.yml]: Rename to reload

File: tasks/main.yml [hunk 6]
```diff
@@ -1,3 +1,21 @@
...
+  template:
+    src: jail.local.j2
+    dest: /etc/fail2ban/jail.local
+  notify: Restart fail2ban
+
+- name: Install fail2ban
+  apt:
+    name: fail2ban
+
+- name: Start fail2ban
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]:
```suggestion
- name: Start and enable fail2ban service
```

File: templates/jail.local.j2 [hunk 7]
```diff
@@ -0,0 +1,231 @@
+# {{ ansible_managed }}
+
+#
+# WARNING: heavily refactored in 0.9.0 release.  Please review and
+#          customize settings for your setup.
+#
+# Changes:  in most of the cases you should not modify this
+#           file, but provide customizations in jail.local file,
```
Comment (SlothOfAnarchy) [inline templates/jail.local.j2]: In general, comments that explain how one would manually write this file can be removed. The correct way to populate it is using the role variables. The file shouldn't be edited by hand and if one needs hints on how to configure it, the role README should give them.

File: templates/jail.local.j2 [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+action = %(action_)s
+
+
+#
+# JAILS
+#
+
+{% for jail in fail2ban_jails %}
+
+[{{ jail.name }}]
```
Comment (SlothOfAnarchy) [inline templates/jail.local.j2]: this config format looks like an ini to me
there is an Ansible module to configure ini files, maybe this could be helpful to use instead of templating

**Review by SlothOfAnarchy — CHANGES_REQUESTED** [review]
I didn't fully review the Jinja template, but I think you're not done yet anyway

**Comment by adriaan-niess** [comment]
All issus except making use of an ini-module should be resolved now.

File: templates/jail.local.j2 [hunk 9]
```diff
@@ -224,8 +224,10 @@ action = %(action_)s
...
 port = {{ jail.port }}
 logpath = {{ jail.logpath }}
-{% for opt_field in ['backend', 'bantime', 'findtime', 'maxretry', 'enabled'] %}
+{% for opt_field in ['backend', 'bantime', 'findtime', 'maxretry'] %}
 {% if jail[opt_field] is defined %}{{ opt_field }} = {{ jail[opt_field] }}{% endif %}
 {% endfor %}
 
+{% if jail['enabled'] is defined %}enabled = {{ jail['enabled'] | bool | lower }}{% endif %}
+
 {% endfor %}
```
Comment (SlothOfAnarchy) [inline templates/jail.local.j2]: no newline at EOF

File: defaults/main.yml (hunk 2 above)
Comment (adriaan-niess) [reply defaults/main.yml:10]: What's the advantage of using a dict? Is this how it's done in other roles?

Comment (haslersn) [reply defaults/main.yml:10]: The advantage is: With a dict you can't accidentally assign a name twice.

File: README.md [hunk 10]
```diff
@@ -1,27 +1,53 @@
...
-A brief description of the role goes here.
-
+This role sets up fail2ban.
 
 ## Requirements
 
-Any pre-requisites that may not be covered by Ansible itself or the role should be mentioned here.
-For instance, if the role uses the EC2 module or depends on other Ansible roles, it may be a good idea to mention in this section that the boto package is required.
-
+No requirements, but can be used in combination with the [nginx](https://github.com/stuvusIT/nginx) role.
```
Comment (SlothOfAnarchy) [inline README.md]:
```suggestion
An apt-based system with `fail2ban` in a configured repository.
The role can be used in combination with the [nginx](https://github.com/stuvusIT/nginx) role. 
```

File: README.md [hunk 11]
```diff
@@ -1,27 +1,53 @@
...
 ## Role Variables
 
-A description of the settable variables for this role should go here, including any variables that are in defaults/main.yml, vars/main.yml, and any variables that can/should be set via parameters to the role.
-Any variables that are read from other roles and/or the global scope (ie. hostvars, group vars, etc.) should be mentioned here as well.
-
-```yml
-```
+### General
+
+| Name | Required/Default | Description |
```
Comment (SlothOfAnarchy) [inline README.md]: Please indent the tables in order to make them readable in a text editor.

File: README.md [hunk 12]
```diff
@@ -1,27 +1,53 @@
...
+|------|------------------|-------------|
+| `fail2ban_bantime` | `3600` | Time to ban hosts. Can be overwritten for individual jails. |
+| `fail2ban_findtime` | `600` | If a hosts exceeds `fail2ban_maxretry` violations within this timeframe it will get banned. Can be overwritten for individual jails. |
+| `fail2ban_maxretry` | `3` | Maximum violations until ban. Can be overwritten for individual jails. |
+| `fail2ban_enabled` | `False` | On default all jails are disabled and must be enabled individually. |
```
Comment (SlothOfAnarchy) [inline README.md]:
```suggestion
| `fail2ban_enabled` | `False` | Whether to enable all jails by default. |
```

Wouldn't `True` be a better default? In most cases, one wants to use the jails he configured.

File: README.md [hunk 13]
```diff
@@ -1,27 +1,53 @@
...
+
+`fail2ban_jails` is a list of objects with the following attributes:
+
+| Name | Required/Default | Description |
+|------|------------------|-------------|
+| `name` | :heavy_check_mark: | Name of the jail. |
+| `port` | :heavy_check_mark: | Portnumber to check. Is only considered by certain filters. |
+| `logpath` | :heavy_check_mark: | Logfile to check. |
+| `backend` | :heavy_multiplication_x: | Backend e.g. `systemd` or `auto` |
+| `enabled` | `False` | Enables/Disables the jail. |
```
Comment (SlothOfAnarchy) [inline README.md]:
```suggestion
| `enabled` | `{{ fail2ban_enabled }}` | Enables/Disables the jail. |
```

File: README.md [hunk 14]
```diff
@@ -1,27 +1,53 @@
...
+### Jails
+
+`fail2ban_jails` is a list of objects with the following attributes:
+
+| Name | Required/Default | Description |
+|------|------------------|-------------|
+| `name` | :heavy_check_mark: | Name of the jail. |
+| `port` | :heavy_check_mark: | Portnumber to check. Is only considered by certain filters. |
+| `logpath` | :heavy_check_mark: | Logfile to check. |
+| `backend` | :heavy_multiplication_x: | Backend e.g. `systemd` or `auto` |
```
Comment (SlothOfAnarchy) [inline README.md]:
```suggestion
| `backend` | `{{ fail2ban_backend }}` | Backend e.g. `systemd` or `auto` |
```

File: README.md [hunk 15]
```diff
@@ -1,27 +1,53 @@
...
+| `enabled` | `False` | Enables/Disables the jail. |
+| `maxretry` | :heavy_multiplication_x: | Maximum violations until ban. |
```
Comment (SlothOfAnarchy) [inline README.md]:
```suggestion
| `maxretry` | `{{ fail2ban_maxretry }}` | Maximum violations until ban. |
```

File: README.md [hunk 16]
```diff
@@ -1,27 +1,53 @@
...
+| `bantime` | :heavy_multiplication_x: | Time to ban hosts. |
+| `findtime` | :heavy_multiplication_x: | If a hosts exceeds `fail2ban_maxretry` or respectively `maxretry` violations within this timeframe it will get banned. |
```
Comment (SlothOfAnarchy) [inline README.md]:
```suggestion
| `findtime` | `{{ fail2ban_findtime }}` | If a hosts exceeds `fail2ban_maxretry` or respectively `maxretry` violations within this timeframe it will get banned. |
```

File: README.md [hunk 17]
```diff
@@ -1,27 +1,53 @@
...
+
+| Name | Required/Default | Description |
+|------|------------------|-------------|
+| `name` | :heavy_check_mark: | Name of the jail. |
+| `port` | :heavy_check_mark: | Portnumber to check. Is only considered by certain filters. |
+| `logpath` | :heavy_check_mark: | Logfile to check. |
+| `backend` | :heavy_multiplication_x: | Backend e.g. `systemd` or `auto` |
+| `enabled` | `False` | Enables/Disables the jail. |
+| `maxretry` | :heavy_multiplication_x: | Maximum violations until ban. |
+| `bantime` | :heavy_multiplication_x: | Time to ban hosts. |
```
Comment (SlothOfAnarchy) [inline README.md]:
```suggestion
| `bantime` | `{{ fail2ban_bantime }}` | Time to ban hosts. |
```

File: README.md [hunk 18]
```diff
@@ -1,27 +1,53 @@
...
+| `fail2ban_ignoreip` | `["127.0.0.1/8"]` | List of IPs that must not get banned. |
+
+### Jails
+
+`fail2ban_jails` is a list of objects with the following attributes:
+
+| Name | Required/Default | Description |
+|------|------------------|-------------|
+| `name` | :heavy_check_mark: | Name of the jail. |
+| `port` | :heavy_check_mark: | Portnumber to check. Is only considered by certain filters. |
```
Comment (SlothOfAnarchy) [inline README.md]:
```suggestion
| `port` | :heavy_check_mark: | Portnumber to check. Is only considered by certain filters. Numbers or names of well-defined ports (e.g. `ssh`, `http`, `sftp`) are allowed. |
```

File: README.md [hunk 19]
```diff
@@ -1,27 +1,53 @@
...
 
 ## Example Playbook
 
-Including an example of how to use your role (for instance, with variables passed in as parameters) is always nice for users too:
-
-```yml
+```
+fail2ban_jails:
+  - name: wordpress
+    port: http,https
```
Comment (SlothOfAnarchy) [inline README.md]: `port` seems to be a string that contains a comma-separated list. I suggest using the ansible list type instead. In the template, you can use `{{ port | join(",") }}`.
We could also define an empty list as default, then we only have to set this key when we actually need it for the jail.

File: defaults/main.yml (hunk 2 above)
Comment (SlothOfAnarchy) [reply defaults/main.yml:10]: Also, you could even define a default configuration for a jail in a group yml and override only some attribute of it in a host yml.

File: defaults/main.yml [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+    port: ssh
+    logpath: "%(sshd_log)s"
+    backend: systemd
+    enabled: True
+  - name: sshd-ddos
+    port: ssh
+    logpath: "%(sshd_log)s"
+    backend: systemd
+    enabled: True
+  - name: nginx-limit-req
```
Comment (SlothOfAnarchy) [inline defaults/main.yml]: Are the jails pre-defined and only selected using the name? Can you reference a list of available jail configurations in the Readme?

File: README.md [hunk 21]
```diff
@@ -1,27 +1,53 @@
 # Role Name
```
Comment (SlothOfAnarchy) [inline README.md]:
```suggestion
# fail2ban
```

File: files/filter.d/wordpress.conf [hunk 22]
```diff
@@ -0,0 +1,23 @@
+# Wordpress filter
```
Comment (SlothOfAnarchy) [inline files/filter.d/wordpress.conf]:
```suggestion
# Wordpress filter (this file is managed by Ansible - do not edit!)
```
It would be useful to be able to configure filters using Ansible as well. If they are in the role, I need to change the role each time, or worse, fork it if I don't have push access.

File: tasks/main.yml [hunk 23]
```diff
@@ -1,3 +1,39 @@
 ---
+- name: Make folder /etc/fail2ban
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]:
```suggestion
- name: Create configuration directory /etc/fail2ban
```

File: templates/jail.local.j2 [hunk 24]
```diff
@@ -0,0 +1,205 @@
...
+#
+# NOTE: This action relies on banaction being present on start and therefore
+# should be last action defined for a jail.
+#
+action_badips = badips.py[category="%(__name__)s", banaction="%(banaction)s", agent="%(fail2ban_agent)s"]
+#
+# Report ban via badips.com (uses action.d/badips.conf for reporting only)
+#
+action_badips_report = badips[category="%(__name__)s", agent="%(fail2ban_agent)s"]
+
```
Comment (SlothOfAnarchy) [inline templates/jail.local.j2:179]: I think it would be useful to be able to configure filters and actions using Ansible variables, but that may be work for another pull request.

**Review by SlothOfAnarchy — CHANGES_REQUESTED** [review]

File: templates/jail.local.j2 (hunk 24 above)
Comment (adriaan-niess) [reply templates/jail.local.j2:179]: Do you mean adding custom files implementing filters and action? Or do you mean assigning filter/actions to jails? Because the second option is already possible.

File: README.md (hunk 19 above)
Comment (adriaan-niess) [reply README.md]: Seems like a good idea.

File: templates/jail.local.j2 (hunk 8 above)
Comment (adriaan-niess) [reply templates/jail.local.j2]: It's indeed an ini file. However @Mr-Pi told me that the ini module comes with some disadvantages and the template module is usually the better choice.

File: defaults/main.yml (hunk 2 above)
Comment (adriaan-niess) [reply defaults/main.yml:10]: Do you mean for the `fail2ban_additional_filters` field (which was removed for now)? Or for the whole config?

File: defaults/main.yml (hunk 20 above)
Comment (AdriaanNiess) [reply defaults/main.yml]: At the moment they aren't predefined. But I'm going to add a list of predefined jails later on. Preferably in another merge request.

Comment (adriaan-niess) [reply defaults/main.yml]: I plan on adding more predefined jails later on in a seperate pull request and also document them in the readme.
Right now the predefined jails are more for testing purposes.

File: defaults/main.yml (hunk 2 above)
Comment (SlothOfAnarchy) [reply defaults/main.yml:10]: I don't see `fail2ban_additional_filters`, I mean `fail2ban_jails`, i.e. the type of `fail2ban_jails` would be a dict with entries such as:
```yml
fail2ban_jails:
  ssh-ddos:
    ports:
      - ssh
    logpath: %(sshd_log)s
    backend: systemd
    enabled: True
```

File: README.md [hunk 25]
```diff
@@ -1,41 +1,42 @@
...
-| `bantime` | :heavy_multiplication_x: | Time to ban hosts. |
-| `findtime` | :heavy_multiplication_x: | If a hosts exceeds `fail2ban_maxretry` or respectively `maxretry` violations within this timeframe it will get banned. |
-| `filters` | :heavy_multiplication_x: | Name of the filter to be applied. |
-| `action` | :heavy_multiplication_x: | Name of the action to be applied if a filter matches. |
+| Name       | Required/Default          | Description                                                                                                                                            |
+|------------|---------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `name`     | :heavy_check_mark:        | Name of the jail.                                                                                                                                      |
+| `port`     | :heavy_check_mark:        | List of port numbers to check. Is only considered by certain filters. Numbers or names of well-defined ports (e.g. `ssh`, `http`, `sftp`) are allowed. |
```
Comment (SlothOfAnarchy) [inline README.md]: Please rename this attribute to `ports`

File: templates/jail.local.j2 (hunk 24 above)
Comment (SlothOfAnarchy) [reply templates/jail.local.j2:179]: I mean creating new filters and actions in a dynamic way, otherwise we have to create a pull request in this role for every new filter/action.
We could an Ansible dict for this, which has the name of the action/filter as a key and a multi-line string as content. The content could then be written into the fail2ban config or into separate files.

File: README.md (hunk 12 above)
Comment (SlothOfAnarchy) [reply README.md]: Wouldn't True be a better default? In most cases, one wants to use the jails he configured.

Comment (adriaan-niess) [reply README.md]: Well that's the default in the generic fail2ban config. There are a lot of predefined jails and you can enable them one by one if you need them. Or alternativley if you want to enable all of them you set the `fail2ban_enabled` variable to `true`.

I think it's better to not enable every jail on default, because jails might require specific configurations depending on the host setup (e.g. reverse proxy) and can potentially have a severe impact on the hosts operation if misconfigured.

File: templates/jail.local.j2 (hunk 24 above)
Comment (adriaan-niess) [reply templates/jail.local.j2:179]: To clarify: the statement
```
action_badips_report = badips[category="%(__name__)s", agent="%(fail2ban_agent)s"]
```
defines a new action `action_badips_report` based on another action `badips` defined in `action.d/badips.conf`. It only sets two variables. Right now redefining actions in such a way isn't possible, but I guess we won't need such a feature anytime soon. It's possible to add that later.

Right now it's possible to specify the actions used in jails. For example a host could be configured like this:

```
fail2ban_jails:
  wordpress:
    filter: wordpress
    action: nginx
    …
```

The `wordpress` jail will use the filter `filter.d/wordpress.conf` and the action `action.d/nginx.conf`. So it's possible to specify what filter/action is used by a jail. However it's not possible atm to add custom actions/filters yet, but I'm going to add such a feature later on in a seperate pull request. It's basically nothing more than a file upload feature.

It might be useful to know that fail2ban already comes with a lot of filters and actions on default (e.g. an action for iptables). So I guess that we don't need to add custom filters and actions that often.

File: README.md [hunk 26]
```diff
@@ -1,27 +1,54 @@
...
+| `fail2ban_findtime`    | `600`             | If a hosts exceeds `fail2ban_maxretry` violations within this timeframe it will get banned. Can be overwritten for individual jails. |
+| `fail2ban_maxretry`    | `3`               | Maximum violations until ban. Can be overwritten for individual jails.                                                               |
+| `fail2ban_enabled`     | `False`           | Whether to enable all jails by default.                                                                                              |
+| `fail2ban_backend`     | `systemd`         | Default backend used to read logs and detect violations.                                                                             |
+| `fail2ban_logencoding` | `auto`            |                                                                                                                                      |
+| `fail2ban_ignoreip`    | `["127.0.0.1/8"]` | List of IPs that must not get banned.                                                                                                |
+
+### Jails
+
+`fail2ban_jails` is a list of objects with the following attributes:
```
Comment (SlothOfAnarchy) [inline README.md]:
```suggestion
`fail2ban_jails` is a dict where the key corresponds to the jail name and the value is an object with the following attributes:
```

File: README.md [hunk 27]
```diff
@@ -1,27 +1,54 @@
...
+
+| Name       | Required/Default          | Description                                                                                                                                            |
+|------------|---------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `name`     | :heavy_check_mark:        | Name of the jail.                                                                                                                                      |
```
Comment (SlothOfAnarchy) [inline README.md]:
```suggestion
```

**Review by SlothOfAnarchy — APPROVED** [review]

**Review by SlothOfAnarchy — APPROVED** [review]
